[UnifiedPDF] Highlighted text in PDF disappears momentarily when Safari is resized and zoomed

aprotyas opened the pull request:
#### 0f33d68c956ec9a92c43603f521795a29559022f
<pre>
[UnifiedPDF] Highlighted text in PDF disappears momentarily when Safari is resized and zoomed
<a href="https://bugs.webkit.org/show_bug.cgi?id=273523">https://bugs.webkit.org/show_bug.cgi?id=273523</a>
<a href="https://rdar.apple.com/125426369">rdar://125426369</a>

Reviewed by Simon Fraser.

Our current selection drawing strategy is to hand off a graphics context
to the `-[PDFSelection drawForPage:withBox:active:inContext:]` API,
following which PDFKit duly manipulates the context with selection
geometry. Unfortunately, PDFKit uses the &quot;multiply&quot; compositing mode when
filling the selection paths in our supplied context. This can result in
the white flashing described in the bug report in cases where WebKit has
not painted anything into the associated layers.

This patch addresses this issue by setting up our own graphics layer in
which to paint PDF selections. To do so, we adopt the
`-[PDFSelection enumerateRectsAndTransformsForPage:usingBlock:]` SPI,
which gives us the selection geometry information necessary to fill the
current path of our graphics context.

* Source/WTF/wtf/PlatformEnableCocoa.h:

Add a compile time flag to guard the new &quot;paint selections into our own
layer&quot; behavior behind platform availability of requisite PDFSelection
SPI.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ColorBlending.h:

Export the blendSourceOver() method, which we use to blend the selection
background color with opaque white, thus creating the final (opaque)
selection appearance.

We also annotate the blendSourceOver declaration with parameter names
for clarity, and leave a cautionary note above the blendWithWhite()
declaration.

* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:

Introduce a new GraphicsLayer owned by the plugin, exclusively for
painting PDF selections. Note that this is guarded behind enablement of
the plugin painting its own selections, which itself depends on platform
availability for the requisite PDFSelection SPI.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setNeedsRepaintInDocumentRect):

Make sure to ask the selection layer to be repainted if a selection is
the repaint reason.

(WebKit::UnifiedPDFPlugin::ensureLayers):
(WebKit::UnifiedPDFPlugin::updateLayerHierarchy):
(WebKit::UnifiedPDFPlugin::updateLayerPositions):
(WebKit::UnifiedPDFPlugin::didChangeSettings):
(WebKit::UnifiedPDFPlugin::layerNeedsPlatformContext const):

Make sure we return `false` under the new behavior where the plugin can
paint selections into a layer it owns. This allows us to enable
accelerated drawing for the selection layer, since it will not be backed
by the WP.

(WebKit::UnifiedPDFPlugin::didChangeIsInWindow):

Propagate isInWindow update to the selection layer such that it paints
correctly as a TiledBacking layer.

(WebKit::UnifiedPDFPlugin::paintContents):
(WebKit::UnifiedPDFPlugin::paintPDFContent):

To not vend selection painting duty to PDFKit, while also maintaining
correct selection painting behavior on platforms that do not have the
new PDFSelection SPI, make sure that haveSelection is invariably false
when we want to paint selections into our own layer.

(WebKit::UnifiedPDFPlugin::paintPDFSelection):

Avoid the problematic &quot;multiply&quot; compositing mode, instead painting
opaquely into the context and compositing via the selection layer, which
is configured with a &quot;multiply&quot; blend mode.

(WebKit::UnifiedPDFPlugin::canPaintSelectionIntoOwnedLayer const):

Returns true iff the requisite PDFSelection interface is available at
runtime.

(WebKit::UnifiedPDFPlugin::determineCurrentlySnappedPage):

Canonical link: <a href="https://commits.webkit.org/278263@main">https://commits.webkit.org/278263@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6de504e416889f603b9470e7a2d3f7bda937db05

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49976 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/29264 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/2259 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/53219 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/653 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/52277 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/35369 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/222 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/40761 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/52074 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42990 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21895 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/24291 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/230 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/8346 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/43299 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/46358 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/278 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54800 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/49471 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/25070 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/207 "Found 1 new test failure: imported/w3c/web-platform-tests/web-locks/query.tentative.https.any.worker.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/48153 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/26328 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/43193 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/47209 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10960 "Build is in progress. Recent messages:OS: Ventura (13.6.6), Xcode: 14.3; Checked out pull request; Reviewed by Simon Fraser; Running compile-webkit") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/27190 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/56954 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7215 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/26057 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/11699 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -450,6 +459,16 @@ static String mutationObserverNotificationString()
+
+#if ENABLE(UNIFIED_PDF_PAINT_SELECTIONS)
+    if (!m_selectionsLayer) {
+        // FIXME: Evaluate whether this should be a TiledBacking layer and hooked up to an async renderer.
```
Yes it should be (it can get as large as the contents). No need to hook up the async renderer.

aprotyas replied:
Addressed in https://github.com/WebKit/WebKit/commit/3175c379f8966936bbf4d9833ad4b1c11e7a1ede

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -450,6 +459,16 @@ static String mutationObserverNotificationString()
+        // FIXME: Evaluate whether this should be a TiledBacking layer and hooked up to an async renderer.
+        m_selectionsLayer = createGraphicsLayer("PDF selections"_s, GraphicsLayer::Type::Normal);
+        m_selectionsLayer->setAnchorPoint({ });
+        m_selectionsLayer->setDrawsContent(true);
```
add `setAcceleratesDrawing(true)`

aprotyas replied:
Addressed in https://github.com/WebKit/WebKit/commit/3175c379f8966936bbf4d9833ad4b1c11e7a1ede

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
```diff
@@ -543,6 +546,9 @@ class UnifiedPDFPlugin final : public PDFPluginBase, public WebCore::GraphicsLay
     RefPtr<WebCore::GraphicsLayer> m_pageBackgroundsContainerLayer;
     RefPtr<WebCore::GraphicsLayer> m_contentsLayer;
+#if ENABLE(UNIFIED_PDF_PAINT_SELECTIONS)
+    RefPtr<WebCore::GraphicsLayer> m_selectionsLayer;
```
`m_selectionLayer`

aprotyas replied:
Addressed in https://github.com/WebKit/WebKit/commit/3175c379f8966936bbf4d9833ad4b1c11e7a1ede

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -960,6 +995,56 @@ static String mutationObserverNotificationString()
+                if (AffineTransform transform { cgTransform }; !transform.isIdentity())
+                    context.concatCTM(transform);
+
+                context.fillRect({ cgRect }, selectionColor, CompositeOperator::SourceOver, BlendMode::Multiply);
```
This is wrong when painting into the selection layer; the layer itself needs to have the Multiply blend mode set, and you should paint opaquely. Overlapping selection rects should not show the overlap.

aprotyas replied:
Addressed in https://github.com/WebKit/WebKit/commit/3175c379f8966936bbf4d9833ad4b1c11e7a1ede

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -960,6 +1013,60 @@ static String mutationObserverNotificationString()
+        if ([m_currentSelection respondsToSelector:@selector(enumerateRectsAndTransformsForPage:usingBlock:)]) {
+            [protectedCurrentSelection() enumerateRectsAndTransformsForPage:page.get() usingBlock:[&context, &selectionColor](CGRect cgRect, CGAffineTransform cgTransform) mutable {
+                // FIXME: Perf optimization -- consider coalescing rects by transform.
+                GraphicsContextStateSaver individualRectTransformPairStateSaver { context };
```
you should avoid this save/restore pair if you don't concatCTM

aprotyas replied:
I changed the state saver to not unconditionally save-and-restore. It only saves before actually concatCTM-ing now. https://github.com/WebKit/WebKit/commit/ab434ecd8a237eb5256d62fbc8744be2db5813fc

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -960,6 +1013,60 @@ static String mutationObserverNotificationString()
+                if (AffineTransform transform { cgTransform }; !transform.isIdentity())
+                    context.concatCTM(transform);
+
+                context.fillRect({ cgRect }, selectionColor, CompositeOperator::SourceOver);
```
you don't need to say SourceOver here

aprotyas replied:
Dropped in https://github.com/WebKit/WebKit/commit/ab434ecd8a237eb5256d62fbc8744be2db5813fc

hortont424 commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -707,6 +707,10 @@
 #endif
 
+#if !defined(ENABLE_UNIFIED_PDF_PAINT_SELECTIONS) && ENABLE(UNIFIED_PDF) && HAVE(PDFSELECTION_ENUMERATE_RECTS_AND_TRANSFORMS)
+#define ENABLE_UNIFIED_PDF_PAINT_SELECTIONS 1
```
maybe UNIFIED_PDF_SELECTION_LAYER?

aprotyas replied:
Renamed in https://github.com/WebKit/WebKit/commit/ab434ecd8a237eb5256d62fbc8744be2db5813fc

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -875,20 +928,20 @@ static String mutationObserverNotificationString()
-    if (m_currentSelection && shouldPaintSelection) {
-        // FIXME: Also test is m_currentSelection is not empty.
+#if !ENABLE(UNIFIED_PDF_PAINT_SELECTIONS)
+    bool shouldPaintSelection = behavior == PaintingBehavior::All;
```
shouldn't `shouldPaintSelection` be the one that takes UNIFIED_PDF_PAINT_SELECTIONS into account?

aprotyas replied:
Indeed, I've changed this logic a little bit and used the helper `canPaintSelectionIntoOwnedLayer()` in https://github.com/WebKit/WebKit/commit/ab434ecd8a237eb5256d62fbc8744be2db5813fc

hortont424 commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -742,6 +773,15 @@ static String mutationObserverNotificationString()
 
+bool UnifiedPDFPlugin::layerNeedsPlatformContext(const WebCore::GraphicsLayer* layer) const
+{
+#if ENABLE(UNIFIED_PDF_PAINT_SELECTIONS)
```
After talking to Simon, it sounds like we no longer need a platform context at all if we're in your new mode, so you should not need the layer check here. Also, the *not* enabled case should be `true`, not false, right? (we need a platform context to paint selections without the new API)

aprotyas replied:
Ah yes, the _not_ enabled case should return `true`.

aprotyas replied:
Okay, since none of the layers are WP-backed, I'll just remove the layer check.

aprotyas replied:
Addressed in https://github.com/WebKit/WebKit/commit/ab434ecd8a237eb5256d62fbc8744be2db5813fc

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -450,6 +460,17 @@ static String mutationObserverNotificationString()
+
+#if ENABLE(UNIFIED_PDF_SELECTION_LAYER)
+    if (!m_selectionLayer) {
+        m_selectionLayer = createGraphicsLayer("PDF selections"_s, isFullMainFramePlugin() ? GraphicsLayer::Type::PageTiledBacking : GraphicsLayer::Type::TiledBacking);
```
Just unconditionally use a `GraphicsLayer::Type::TiledBacking`.

aprotyas replied:
Addressed in https://github.com/WebKit/WebKit/commit/6de504e416889f603b9470e7a2d3f7bda937db05

aprotyas committed:
[UnifiedPDF] Highlighted text in PDF disappears momentarily when Safari is resized and zoomed
https://bugs.webkit.org/show_bug.cgi?id=273523
rdar://125426369

Reviewed by Simon Fraser.

Our current selection drawing strategy is to hand off a graphics context
to the `-[PDFSelection drawForPage:withBox:active:inContext:]` API,
following which PDFKit duly manipulates the context with selection
geometry. Unfortunately, PDFKit uses the "multiply" compositing mode when
filling the selection paths in our supplied context. This can result in
the white flashing described in the bug report in cases where WebKit has
not painted anything into the associated layers.

This patch addresses this issue by setting up our own graphics layer in
which to paint PDF selections. To do so, we adopt the
`-[PDFSelection enumerateRectsAndTransformsForPage:usingBlock:]` SPI,
which gives us the selection geometry information necessary to fill the
current path of our graphics context.

* Source/WTF/wtf/PlatformEnableCocoa.h:

Add a compile time flag to guard the new "paint selections into our own
layer" behavior behind platform availability of requisite PDFSelection
SPI.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/graphics/ColorBlending.h:

Export the blendSourceOver() method, which we use to blend the selection
background color with opaque white, thus creating the final (opaque)
selection appearance.

We also annotate the blendSourceOver declaration with parameter names
for clarity, and leave a cautionary note above the blendWithWhite()
declaration.

* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/rendering/RenderTheme.h:
* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:

Introduce a new GraphicsLayer owned by the plugin, exclusively for
painting PDF selections. Note that this is guarded behind enablement of
the plugin painting its own selections, which itself depends on platform
availability for the requisite PDFSelection SPI.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::setNeedsRepaintInDocumentRect):

Make sure to ask the selection layer to be repainted if a selection is
the repaint reason.

(WebKit::UnifiedPDFPlugin::ensureLayers):
(WebKit::UnifiedPDFPlugin::updateLayerHierarchy):
(WebKit::UnifiedPDFPlugin::updateLayerPositions):
(WebKit::UnifiedPDFPlugin::didChangeSettings):
(WebKit::UnifiedPDFPlugin::layerNeedsPlatformContext const):

Make sure we return `false` under the new behavior where the plugin can
paint selections into a layer it owns. This allows us to enable
accelerated drawing for the selection layer, since it will not be backed
by the WP.

(WebKit::UnifiedPDFPlugin::didChangeIsInWindow):

Propagate isInWindow update to the selection layer such that it paints
correctly as a TiledBacking layer.

(WebKit::UnifiedPDFPlugin::paintContents):
(WebKit::UnifiedPDFPlugin::paintPDFContent):

To not vend selection painting duty to PDFKit, while also maintaining
correct selection painting behavior on platforms that do not have the
new PDFSelection SPI, make sure that haveSelection is invariably false
when we want to paint selections into our own layer.

(WebKit::UnifiedPDFPlugin::paintPDFSelection):

Avoid the problematic "multiply" compositing mode, instead painting
opaquely into the context and compositing via the selection layer, which
is configured with a "multiply" blend mode.

(WebKit::UnifiedPDFPlugin::canPaintSelectionIntoOwnedLayer const):

Returns true iff the requisite PDFSelection interface is available at
runtime.

(WebKit::UnifiedPDFPlugin::determineCurrentlySnappedPage):

Canonical link: https://commits.webkit.org/278263@main
